devicedefender/include/aws/iotdevicedefender/DeviceDefender.h

```diff
@@ -174,56 +287,44 @@ namespace Aws
 
             static void s_onDefenderV1TaskCancelled(void *userData);
 
-            struct customMetricData
-            {
-                size_t index;
-                ReportTask *task;
-            };
-            static int s_getCustomMetricNumber(double *output, void *customData);
-            static int s_getCustomMetricNumberList(aws_array_list *output, void *customData);
-            static int s_getCustomMetricStringList(aws_array_list *output, void *customData);
-            static int s_getCustomMetricIpList(aws_array_list *output, void *customData);
-            std::vector<std::function<int(double *)>> m_storedCustomMetricsNumberFunctions;
-            std::vector<std::function<int(std::vector<double> *)>> m_storedCustomMetricsNumberListFunctions;
-            std::vector<std::function<int(std::vector<std::string> *)>> m_storedCustomMetricsStringListFunctions;
-            std::vector<std::function<int(std::vector<std::string> *)>> m_storedCustomMetricsIpListFunctions;
-
-            std::vector<customMetricData *> storedCustomMetricData = std::vector<customMetricData *>();
+            // Holds all of the custom metrics created for this task. These are pointers that will be
+            // automatically created and cleaned by ReportTask when it is destroyed.
+            Crt::Vector<CustomMetricBase *> storedCustomMetrics;
```

Comment:
Use shared_ptr or unique_ptr.  I'd use shared_ptr even if not necessary as then the deleter can be hidden from the type and you shouldn't need to keep an allocator member either.